src/ddmd/dinterpret.d

```diff
@@ -2047,6 +2047,23 @@ public:
             assert(result.op == TOKstructliteral || result.op == TOKclassreference);
             return;
         }
+
+        // we probably have a nested delegate here.
+        // the thing to do is to iterate up the stack until we find
+        // a this that has a matching type.
+        // https://issues.dlang.org/showbug.cgi?id=16301
+        size_t stackOffset = ctfeStack.savedThis.dim;
+        while(stackOffset && (!result || !result.type.equals(e.type)))
```

Comment:
That's right, ctfeStack is the call stack.
However if a there is a closure it must have been created by a function call, therefore
The aggregate's this must be found and will be the same.
If you look at e2ir's closure handling, it's doing the same thing.
Lexical scope does not matter it just happens to be the same.